### PR TITLE
koord-descheduler: enhance LowNodeLoad scorer

### PR DESCRIPTION
**Pull request by LY-today**
### Ⅰ. Describe what this PR does

What is your proposal:
When descheduler evaluates which pod to expel to reduce the utilization of high-load nodes to a reasonable level, can descheduler perform matching calculations on the actual usage of the pod and the amount of eviction resources in advance?

Why is this needed:
This can reduce the number of times a single node is repeatedly entered into the eviction logical unit, reduce the number of single node eviction instances, and indirectly improve the accuracy of the eviction results and the stability of the business while ensuring that hot issues are solved.

Is there a suggested solution, if so, please add it:
evict pod priority selector：podNowResourceUsages >= nodeNowResourceUsages - nodeHighResourceThresholds * nodeResourceAllocatable

### Ⅱ. Does this pull request fix one issue?
issue链接：https://github.com/koordinator-sh/koordinator/issues/1975

### Ⅲ. Describe how to verify it
I added unit tests

### Ⅳ. Special notes for reviews
none

### V. Checklist

✅ I have written necessary docs and comments
✅ I have added necessary unit tests and integration tests
✅ All checks passed in `make test`

### CODE REVIEW

**Comment by LY-today** [comment]
@eahydra 你好，E2E-k8s-1.22 中日志显示：I0608 08:49:27.601454       1 leaderelection.go:285] failed to renew lease koordinator-system/koordinator-manager: timed out waiting for the condition

似乎这个错误并不是我引起的，请问有什么解决办法吗？

**Comment by LY-today** [comment]
/assign @eahydra Hi, eahydra, E2E-k8s-1.22 log shows: I0608 08:49:27.601454 1 leaderelection.go:285] failed to renew lease koordinator-system/koordinator-manager: timed out waiting for the condition

It seems that this error is not caused by me. Is there any solution?

**Comment by LY-today** [comment]
@saintube hi,The tide process seems to be stuck, what information do I need to add to move the process forward?

**Comment by songchh** [comment]
Agreed, eviction based solely on the maximum usage of pods will indeed lead to stability problems. The simplest scenario is that after the pod with the largest usage is evicted, few nodes in the scheduler can meet its resource usage requirements, and pending becomes the norm. Even if it can be scheduled on certain nodes, the single-machine hotspot problem is more likely to be triggered during peak usage of multiple pods. Therefore, I agree that this should be addressed by evicting pods whose usage is just enough to meet demand, rather than simply based on maximum usage.

**Comment by LY-today** [comment]
@ZiMengSheng hi,The tide process seems to be stuck, what information do I need to add to move the process forward?
@songtao98 hi,The tide process seems to be stuck, what information do I need to add to move the process forward?

**Comment by LY-today** [comment]
@saintube @eahydra Could you please arrange for a classmate to follow up on this PR? It is currently affecting the stability of our business on the internal cluster. Thank you all.

**Comment by songtao98** [comment]
> @saintube @eahydra Could you please arrange for a classmate to follow up on this PR? It is currently affecting the stability of our business on the internal cluster. Thank you all.

@LY-today I'm reviewing this and sorry for the delay. Plesae wait a little bit, I'll do my best and review it ASAP.

**Comment by LY-today** [comment]
> > @ZiMengSheng hi,The tide process seems to be stuck, what information do I need to add to move the process forward? @songtao98 hi,The tide process seems to be stuck, what information do I need to add to move the process forward?
> 
> No more other things to do for now, just need a review and approval.
> 
> > @saintube @eahydra Could you please arrange for a classmate to follow up on this PR? It is currently affecting the stability of our business on the internal cluster. Thank you all.
> 
> @LY-today I'm reviewing this and sorry for the delay. Plesae wait a little bit, I'll do my best and review it ASAP.


thank you for your reply

**Comment by LY-today** [comment]
> Agreed, eviction based solely on the maximum usage of pods will indeed lead to stability problems. The simplest scenario is that after the pod with the largest usage is evicted, few nodes in the scheduler can meet its resource usage requirements, and pending becomes the norm. Even if it can be scheduled on certain nodes, the single-machine hotspot problem is more likely to be triggered during peak usage of multiple pods. Therefore, I agree that this should be addressed by evicting pods whose usage is just enough to meet demand, rather than simply based on maximum usage.


Yes, we seem to have the same problem scenario

File: pkg/descheduler/utils/sorter/scorer.go
Comment (songtao98) [inline pkg/descheduler/utils/sorter/scorer.go:79]: What is the meaning of **negative** here? I think if a pod's `requested` resource is lower than `capacity`, we can just give it a small score but not a negative one.

**Review by songtao98 — COMMENTED** [review]
The basic logic looks good to me. However, let's figure out how to cooperate with PR #2066. The score function need to be discussed.

File: pkg/descheduler/utils/sorter/scorer.go
Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go:79]: > What is the meaning of **negative** here? I think if a pod's `requested` resource is lower than `capacity`, we can just give it a small score but not a negative one.

Negative numbers mean that the score output when ratio is greater than or equal to 1 must be greater than the score output when ratio is less than 1. I tried to think about how to give a positive score when requested is less than capacity, and the score is smaller than the score when requested is greater than or equal to capacity, but unfortunately, no conclusion was reached, so using plural numbers was the fastest and most effective solution I thought at the time.

File: pkg/descheduler/framework/plugins/loadaware/utilization_util.go
Comment (zwForrest) [reply pkg/descheduler/framework/plugins/loadaware/utilization_util.go:477]: Yes, it's essential to sort based on the type of node with high utilization.

File: pkg/descheduler/utils/sorter/scorer.go
Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go:79]: @songtao98 Do you think negative numbers affect code reading or standardization, so do you suggest modifications? Because negative numbers do meet the needs in this scenario, I don't know the main reason why I can suggest modifications.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go:79]: @songtao98 In other words, use positive values ​​for all scores in this scenario. It is difficult to guarantee that the score output when requested>=capacity must be greater than the score output when requested<capacity. If you have a better way to do it, I'd love to hear it.

Comment (songtao98) [reply pkg/descheduler/utils/sorter/scorer.go:79]: > @songtao98 Do you think that negative numbers affect code reading or do not comply with specifications, so do you suggest modifications? Because negative numbers do meet the functional requirements in this scenario, I don't know the main reason for your suggested modification.

I get this. There is only one possible but extreme bad case: Because when we calculate the score of one pod, we calculate through all resources. That means, if a pod get a lower score on one resource(CPU), it could get a higher **final** score than other pods because it get higher scores on **all other** resources(Memory, etc.).  If you make it negative, I think this pod will totally lost a chance to win back. This is why I suggest we find a positive but small score when ratio is < 1 **if possible**. 

An ugly solution is to use combined `piecewise functions. ` For example, 

```
if ratio >= 1 {
		return 1 / ratio + 1 // score is in (1, +∞)
	} else {
		return  ratio // score is in [0, 1)
```

Anyway, I'm doubting this, too. And I'm sure to be open to hear your opinion about whether this is over-designed. :)

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go:79]: @songtao98 After seeing your description, I understand the reason. Indeed, your consideration is more comprehensive and reasonable. I will resubmit using the piecewise functions. Thank you for your suggestion. I have benefited a lot.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: @songtao98 
Submitted, please re-evaluate

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: @songtao98 Is there anything else that needs to be corrected?

Comment (songtao98) [reply pkg/descheduler/utils/sorter/scorer.go]: > @songtao98 Is there anything else that needs to be corrected?

Not for now, this PR looks good to me. But as I said in another comment, there might be a problem with another PR. We will work as soon as possible to finish review and merge it. After that, you might need to modify the way you Sub the resource from the threshold.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: > > @songtao98 Is there anything else that needs to be corrected?
> 
> Not for now, this PR looks good to me. But as I said in another comment, there might be a problem with another PR. We will work as soon as possible to finish review and merge it. After that, you might need to modify the way you Sub the resource from the threshold.

It means that after merging the PR of https://github.com/koordinator-sh/koordinator/pull/2066, this PR needs to be adapted before it can be passed, right?

Comment (songtao98) [reply pkg/descheduler/utils/sorter/scorer.go]: > > > @songtao98 Is there anything else that needs to be corrected?
> > 
> > 
> > Not for now, this PR looks good to me. But as I said in another comment, there might be a problem with another PR. We will work as soon as possible to finish review and merge it. After that, you might need to modify the way you Sub the resource from the threshold.
> 
> It means that after merging the PR of #2066, this PR needs to be adapted before it can be passed, right?

Yes it is. I'll help with your adaption anytime you need. And I'll try to push this process to be quick.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: > > > > @songtao98 Is there anything else that needs to be corrected?
> > > 
> > > 
> > > Not for now, this PR looks good to me. But as I said in another comment, there might be a problem with another PR. We will work as soon as possible to finish review and merge it. After that, you might need to modify the way you Sub the resource from the threshold.
> > 
> > 
> > It means that after merging the PR of #2066, this PR needs to be adapted before it can be passed, right?
> 
> Yes it is. I'll help with your adaption anytime you need. And I'll try to push this process to be quick.

I will listen to your arrangements and thank you for your guidance.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: @songtao98 The latest main branch was merged, and the E2E test failed.

Comment (songtao98) [reply pkg/descheduler/utils/sorter/scorer.go]: > @songtao98 The latest main branch was merged, and the E2E test failed.

will be fixed by #2101

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: > > @songtao98 The latest main branch was merged, and the E2E test failed.
> 
> will be fixed by #2101

”leader election lost“，
I'm afraid you need to re-execute E2E here.

Comment (songtao98) [reply pkg/descheduler/utils/sorter/scorer.go]: > > > @songtao98 The latest main branch was merged, and the E2E test failed.
> > 
> > 
> > will be fixed by #2101
> 
> ”leader election lost“， I'm afraid you need to re-execute E2E here.

Sorry for the late reply. #2066 is now merged into main. You can now rebase main and modify your PR to adapt to the prodThreshold. After that your push should auto execute E2E here.

Comment (LY-today) [reply pkg/descheduler/utils/sorter/scorer.go]: > > > > @songtao98 The latest main branch was merged, and the E2E test failed.
> > > 
> > > 
> > > will be fixed by #2101
> > 
> > 
> > ”leader election lost“， I'm afraid you need to re-execute E2E here.
> 
> Sorry for the late reply. #2066 is now merged into main. You can now rebase main and modify your PR to adapt to the prodThreshold. After that your push should auto execute E2E here.

What exactly should I do? I don’t know the implementation of this PR: https://github.com/koordinator-sh/koordinator/pull/2066

**Comment by songtao98** [comment]
@LY-today Please check and resolve my review comment here: https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998. 

To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`. 

When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.

I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify. 

If any further problems, feel free to comment me. :)

**Comment by songtao98** [comment]
@LY-today Please check and resolve my review comment here: https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998. 

To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`. 

When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.

I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify. 

If any further problems, feel free to comment me. :)

**Comment by songtao98** [comment]
@LY-today Please check and resolve my review comment here: https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998. 

To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`. 

When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.

I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify. 

If any further problems, feel free to comment me. :)

> @LY-today Please check and resolve my review comment here: [#2092 (comment)](https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998).
> 
> To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`.
> 
> When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.
> 
> I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify.
> 
> If any further problems, feel free to comment me. :)

and also, remember to check and modify related code carefully. I'll re-review this PR after your modification. About the failed checking of lint and unit-test, I'm fixing it in #2103 . 

After all, remember to rebase all commits in your PR and just keep only 1 commit.

**Comment by LY-today** [comment]
> @LY-today Please check and resolve my review comment here: [#2092 (comment)](https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998).
> 
> To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`.
> 
> When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.
> 
> I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify.
> 
> If any further problems, feel free to comment me. :)
> 
> > @LY-today Please check and resolve my review comment here: [#2092 (comment)](https://github.com/koordinator-sh/koordinator/pull/2092#discussion_r1635799998).
> > To be short, `prodLowThreshold` and `prodHighThreshold` is just like `lowThreshold` and `highThreshold` but only counting the usage of Prod pods. This is used to balance the prod pods distribution. The newest logic is first balance pods on nodes that are overutilized by `highThreshold` and then balance pods on nodes that are just overutilized by `prodHighThreshold`.
> > When we are doing this, the function `sortPodsOnOneOverloadedNode` is called in function `balancePods`. `balancePods` has an arg `prod bool` and is called first as `balancePods(prod: false)` then `balancePods(prod: true)`. You can see the related code lines.
> > I suggest you add an arg `prod bool` in `sortPodsOnOneOverloadedNode` too. And pass the value of `prod` of `balancePods` to it. If prod is true, you should call `usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold[or])` in `sortPodsOnOneOverloadedNode`. This what I need you to modify.
> > If any further problems, feel free to comment me. :)
> 
> and also, remember to check and modify related code carefully. I'll re-review this PR after your modification. About the failed checking of lint and unit-test, I'm fixing it in #2103 .
> 
> After all, remember to rebase all commits in your PR and just keep only 1 commit.


Received, I am now compatible

**Comment by LY-today** [comment]
@songtao98 Local make test, the feedback is successful, but the CI link is abnormal. Can it be re-executed manually?

**Comment by songtao98** [comment]
> @songtao98 Local make test, the feedback is successful, but the CI link is abnormal. Can it be re-executed manually?

I noticed that was caused by lint check. You can run `make test`,` make lint`, etc. on your local env to make sure that your commit can pass the checks. Check the make file and you can see what we need.

File: pkg/descheduler/framework/plugins/loadaware/utilization_util.go
Comment (songtao98) [inline pkg/descheduler/framework/plugins/loadaware/utilization_util.go:682]: Here need to be modified as if prod{ usedCopy.Sub(*srcNode.thresholds.prodHighResourceThreshold)}

**Review by songtao98 — CHANGES_REQUESTED** [review]

**Comment by songtao98** [comment]
There is a typo in your PR's title. Consider modify it as koord-descheduler: enhance LowNodeLoad scorer

**Comment by LY-today** [comment]
> @LY-today There is a typo in your PR's title. Consider modify it as koord-descheduler: enhance LowNodeLoad scorer

done

File: pkg/descheduler/framework/plugins/loadaware/utilization_util.go
Comment (LY-today) [reply pkg/descheduler/framework/plugins/loadaware/utilization_util.go:682]: ok

**Comment by LY-today** [comment]
@songtao98 All have been modified and passed the CI stage, please check

**Comment by songtao98** [comment]
/lgtm
@LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.

**Comment by LY-today** [comment]
> /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.

How do I mark it as resolved?

**Comment by LY-today** [comment]
> > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> 
> How do I mark it as resolved?

@songtao98 done

**Comment by LY-today** [comment]
> > > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> > 
> > 
> > How do I mark it as resolved?
> 
> @songtao98 done

@songtao98 What other links do we need to continue to promote the progress of PR integration?

**Comment by songtao98** [comment]
> > > > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> > > 
> > > 
> > > How do I mark it as resolved?
> > 
> > 
> > @songtao98 done
> 
> @songtao98 What other links do we need to continue to promote the progress of PR integration?

An approved label needed. Maintainers will soon do final check ASAP and if no other problems, this PR will be approved and auto-merged.

**Comment by LY-today** [comment]
> > > > > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> > > > 
> > > > 
> > > > How do I mark it as resolved?
> > > 
> > > 
> > > @songtao98 done
> > 
> > 
> > @songtao98 What other links do we need to continue to promote the progress of PR integration?
> 
> An approved label needed. Maintainers will soon do final check ASAP and if no other problems, this PR will be approved and auto-merged.


Received, please work hard and push forward as soon as possible, thank you

**Comment by LY-today** [comment]
> > > > > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> > > > 
> > > > 
> > > > How do I mark it as resolved?
> > > 
> > > 
> > > @songtao98 done
> > 
> > 
> > @songtao98 What other links do we need to continue to promote the progress of PR integration?
> 
> An approved label needed. Maintainers will soon do final check ASAP and if no other problems, this PR will be approved and auto-merged.

Received, thank you for your hard work. We hope to speed up the progress and solve the online problems of our cluster.

**Comment by LY-today** [comment]
> > > > > > /lgtm @LY-today You can mark all conversations as resolved. Then we can continue to merge this PR.
> > > > > 
> > > > > 
> > > > > How do I mark it as resolved?
> > > > 
> > > > 
> > > > @songtao98 done
> > > 
> > > 
> > > @songtao98 What other links do we need to continue to promote the progress of PR integration?
> > 
> > 
> > An approved label needed. Maintainers will soon do final check ASAP and if no other problems, this PR will be approved and auto-merged.
> 
> Received, thank you for your hard work. We hope to speed up the progress and solve the online problems of our cluster.

@songtao98 Hello, are there any plans to advance this PR today?

**Comment by songtao98** [comment]
PTAL @hormes

**Comment by LY-today** [comment]
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[ZiMengSheng](https://github.com/koordinator-sh/koordinator/pull/2092#issuecomment-2172364094)_ To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **eahydra** after the PR has been reviewed.
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=koordinator-sh%2Fkoordinator).
> 
> Needs approval from an approver in each of these files:
> * **[pkg/descheduler/OWNERS](https://github.com/koordinator-sh/koordinator/blob/main/pkg/descheduler/OWNERS)**
> 
> Approvers can indicate their approval by writing `/approve` in a comment Approvers can cancel approval by writing `/approve cancel` in a comment

@eahydra Thank you for your hard work, please review

**Comment by LY-today** [comment]
> [APPROVALNOTIFIER] This PR is **NOT APPROVED**
> 
> This pull-request has been approved by: _[ZiMengSheng](https://github.com/koordinator-sh/koordinator/pull/2092#issuecomment-2172364094)_ To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **eahydra** after the PR has been reviewed.
> 
> The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=koordinator-sh%2Fkoordinator).
> 
> Needs approval from an approver in each of these files:
> * **[pkg/descheduler/OWNERS](https://github.com/koordinator-sh/koordinator/blob/main/pkg/descheduler/OWNERS)**
> 
> Approvers can indicate their approval by writing `/approve` in a comment Approvers can cancel approval by writing `/approve cancel` in a comment

@hormes Please check out this PR and let's work together to solve the problem

**Review by hormes — APPROVED** [review]

**Pull request merged**
